### ISSUE TITLE
[UK] Servers using Swap, Slow Responses. Failing Orders

### ISSUE THREAD

**Issue by lin-d-hop** [openfoodfoundation/openfoodnetwork#7058]
## Description

Multiple reports from multiple users and enterprises that orders are failing because of slow server response times.

https://app.datadoghq.com/dash/host/965502139?from_ts=1614708659057&live=true&to_ts=1615313459057

## Expected Behavior

Successful orders. Reasonable response times. Minimal use of swap

## Actual Behaviour

Failing orders - users redirected to homepage with error informing them to tell us.
Response times are insane if someone uses a report

## Steps to Reproduce

1. Go to OFN UK
2. Do something
3. Wait
4. Wait
5. Wait
6. Wait
7. Leave

## Severity

bug-s1: a critical feature is broken: checkout, payments, signup, login
We're losing orders. This is s1.

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

## Possible Fix

Possible fixes to implement asap please:

1) Turn on Customer Balances performance feature for all users
2) Throw more metal at it. If we need to scale the server for 2 months let's do it.
3) Reduce the timeouts by either 1 or 2 mins please. We can support users to access the stupid reports.

**Comment by Matt-Yorkley** [comment]
We can provision https://github.com/openfoodfoundation/ofn-install/pull/719 and restart the Rails server to toggle customer balances too :+1:

**Comment by sauloperez** [comment]
I think it wouldn't harm to upscale the VPS further and see where more RAM gets us.

**Comment by lin-d-hop** [comment]
We had another pretty severe spike between 19:45 and 21:00 last night.

**Comment by Matt-Yorkley** [comment]
We've got ~10k entries in the nginx logs between 19:15 and 21:15, so I guess it was a busy night. Ordering by highest response time shows the top ~30 worst responses are all the checkout.

**Comment by sauloperez** [comment]
so it may look like the bottleneck are the write operations. We need to investigate that further; from the top, the slow endpoints, to the bottom, PostgreSQL performance. We have the shared gut feeling that this may be related to PostgreSQL abusing disk, because of swap or something else.

**Comment by sauloperez** [comment]
I just run some manual VACUUMs on the most bloated tables, which should make it easier for PostgreSQL to read/write to those tables.

```sql
openfoodnetwork=>  SELECT schemaname, relname, n_live_tup, n_dead_tup, last_autovacuum
FROM pg_stat_all_tables
ORDER BY n_dead_tup / (
  t8n_live_tup * current_setting('autovacuum_vacuum_scale_factor')::float8 + current_setting('autovacuum_vacuum_threshold')::float
) DESC LIMIT 10;                                                                 
 schemaname |           relname            | n_live_tup | n_dead_tup |        last_autovacuum                                   
------------+------------------------------+------------+------------+-------------------------------                           
 pg_catalog | pg_attribute                 |          6 |        132 |                                                          
 pg_catalog | pg_depend                    |          3 |         53 |                                                          
 public     | taggings                     |       2435 |        553 |                                                          
 public     | spree_inventory_units        |    1142836 |     213777 |                                                          
 public     | distributors_payment_methods |        686 |        169 |                                                          
 public     | spree_orders                 |     315175 |      55054 | 2021-03-02 19:53:47.909817+00                            
 public     | sessions                     |     412047 |      71822 | 2021-03-01 04:44:23.330844+00                            
 public     | subscription_line_items      |        544 |        138 |                                                          
 public     | spree_product_properties     |       3723 |        683 |                                                          
 public     | enterprise_roles             |       4255 |        743 |                                                          
(10 rows)   
```
**Results**
<details>
<summary>
openfoodnetwork=> VACUUM (VERBOSE) spree_inventory_units;
</summary>
```sql
INFO:  vacuuming "public.spree_inventory_units"                                                                                 
INFO:  scanned index "spree_inventory_units_pkey" to remove 158053 row versions                                                 
DETAIL:  CPU 0.09s/0.15u sec elapsed 0.58 sec                                                                                   
INFO:  scanned index "index_inventory_units_on_order_id" to remove 158053 row versions
DETAIL:  CPU 0.20s/0.22u sec elapsed 2.32 sec                                                                                   
INFO:  scanned index "index_inventory_units_on_shipment_id" to remove 158053 row versions                                       
DETAIL:  CPU 0.07s/0.17u sec elapsed 0.56 sec
INFO:  scanned index "index_inventory_units_on_variant_id" to remove 158053 row versions                                                                                                                                                                        
DETAIL:  CPU 0.12s/0.32u sec elapsed 0.85 sec  
INFO:  "spree_inventory_units": removed 158053 row versions in 10616 pages
DETAIL:  CPU 0.04s/0.13u sec elapsed 0.31 sec                                                                                                                                                                                                                   
INFO:  index "spree_inventory_units_pkey" now contains 1142836 row versions in 4918 pages
DETAIL:  157462 index row versions were removed.
1 index pages have been deleted, 0 are currently reusable.
CPU 0.00s/0.00u sec elapsed 0.00 sec.
INFO:  index "index_inventory_units_on_order_id" now contains 1142836 row versions in 6878 pages
DETAIL:  158038 index row versions were removed.
10 index pages have been deleted, 9 are currently reusable. 
CPU 0.00s/0.00u sec elapsed 0.00 sec.
INFO:  index "index_inventory_units_on_shipment_id" now contains 1142836 row versions in 5545 pages                         
DETAIL:  135798 index row versions were removed.
3 index pages have been deleted, 3 are currently reusable.  
CPU 0.00s/0.00u sec elapsed 0.00 sec.
INFO:  index "index_inventory_units_on_variant_id" now contains 1142836 row versions in 5881 pages
DETAIL:  123048 index row versions were removed.
450 index pages have been deleted, 450 are currently reusable.
CPU 0.00s/0.00u sec elapsed 0.00 sec.
INFO:  "spree_inventory_units": found 67347 removable, 934480 nonremovable row versions in 13881 out of 16029 pages
DETAIL:  0 dead row versions cannot be removed yet.
There were 31114 unused item pointers.
Skipped 0 pages due to buffer pins.
0 pages are entirely empty.
CPU 0.56s/1.22u sec elapsed 4.92 sec.
VACUUM
```
</details>

**Comment by lin-d-hop** [comment]
Thanks everyone for the excellent team work on this!

One more suggestion before we close this S1 and see how things go over the coming days...

Can we remove old data from:
- Old carts (spree_orders where state = cart)
- spree_state_changes
- sessions

Removing old carts and associated objects has been done in the past:
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/tasks/data/remove_transient_data.rb#L16-L20 

@sauloperez suggests that we:
- review it
- run it on a staging with old data (UK's)
- run it on Katuma (optional)
- run it on UK

Ideally we'll set up a task to do this regularly (once every 3 months) however if that is out of scope for this s1 I'll create another issue.

Thank you :pray:

**Comment by RachL** [comment]
@lin-d-hop Does this delay the last release we want to do before merging Rails? In that case, there is no rush to draft a release today? ping @andrewpbrett

**Comment by lin-d-hop** [comment]
I don't think this should delay the release :pray: 

I know I'm taking a lot of capacity with this... but if a few tweaks can buy us a few months then I think it is time well spent. If you would prefer I'll make the above into a separate issue and we can wait to see if this release makes any impact before trying.

**Comment by RachL** [comment]
@lin-d-hop so it would look like this, please tell me if I understood correctly:

- (hopefully) today UK is upgraded with latests performance improvement

- Rails 5 is merge today or tomorrow

- The list of performance improvement listed here cannot made it to UK prod before Rails 5 roll-out is complete

Is that correct?

**Comment by lin-d-hop** [comment]
Ahhh I understand you were talking about the Rails 5 release now, rather than our mini release today.

I believe that the `Remove Transient Data` task does not need a release plan because it can be run as a once off from the Rails app. As a scheduled task run once every 3 months, I'm not sure how this is implemented, perhaps a dev can help explain.

But in summary - let's not block Rails 5 merge. If things get critical again we can try deploying the Rails5 release to UK before AU and FR as that should include a few performance improvements too :-)

**Comment by andrewpbrett** [comment]
Okay I think we've met the criteria for closing this. The more performant customer balances are switched on, `bundle exec rake ofn:data:remove_transient_data` is run, and #7036 is merged and will be deployed once testing is complete. 

I think we should sort out automating the data cleanup tasks in a separate issue.
